### PR TITLE
Support asynchronous validation rules

### PR DESCRIPTION
**Pull request by Shane32**
Changes `INodeVisitor` to have asynchronous signatures.  Does not change `MatchingNodeVisitor` or `IValidationRule`, so almost all validation rules are unaffected.

This is needed by the server project's authorization validation rule, as ASP.NET Core's authorization framework only provides asynchronous methods to validate a policy rule.

The functionality builds upon the asynchronous node visitor already present within GraphQL.Parser

Note that in a later PR we can add support to `MatchingNodeVisitor` to support asynchronous delegates as a non-breaking change -- if desired.  But users that need asynchronous validation rules can implement `INodeVisitor` directly.

Note that ValidationContext already contains the CancellationToken.

### CODE REVIEW

File: docs2/site/docs/migrations/migration6.md [hunk 1]
```diff
@@ -9,9 +9,18 @@ See [issues](https://github.com/graphql-dotnet/graphql-dotnet/issues?q=milestone
 Especially noteworthy when a data loader is configured with caching enabled and a singleton lifetime,
 memory usage is reduced by freeing unnecessary references after obtaining the results.
 
+### 2. Async support for validation rules
+
+Particularly useful for authentication checks, now validation rules are asynchronous.
```
Comment (sungam3r) [inline docs2/site/docs/migrations/migration6.md:14]: 👍

File: docs2/site/docs/migrations/migration6.md [hunk 2]
```diff
@@ -9,9 +9,18 @@ See [issues](https://github.com/graphql-dotnet/graphql-dotnet/issues?q=milestone
...
 ## Breaking Changes
 
 ### 1. `DataLoaderPair<TKey, T>.Loader` property removed
 
 This property was not used internally and should not be necessary by user code or custom implementations.
 Removal was necessary as the value is released after the result is set.
+
+### 2. `INodeVisitor` and `IVariableVisitor` members' signatures are asynchronous and end in `Async`.
+
+Note that `MatchingNodeVisitor` has not changed, so many validation rules will not require
```
Comment (sungam3r) [inline docs2/site/docs/migrations/migration6.md:25]: `MatchingNodeVisitor` was changed 😕

File: src/GraphQL/Validation/ValidationContext.cs [hunk 3]
```diff
@@ -213,7 +213,7 @@ public Variables GetVariableValues(IVariableVisitor? visitor = null)
                 if (type is IInputObjectGraphType inputObjectGraphType)
                 {
                     var parsedValue = ParseValueObject(inputObjectGraphType, variableDef, variableName, value, visitor);
-                    visitor?.VisitObject(this, variableDef, variableName, inputObjectGraphType, value, parsedValue);
+                    visitor?.VisitObjectAsync(this, variableDef, variableName, inputObjectGraphType, value, parsedValue);
```
[the remark this answers is not in the document]
Comment (Shane32) [reply src/GraphQL/Validation/ValidationContext.cs:216]: Whoops.  I'll probably revert the changes to IVariableVisitor in this PR.  Maybe a separate PR...

[the remark this answers is not in the document]
Comment (Shane32) [reply src/GraphQL/Validation/ValidationContext.cs]: #3166

**Review by sungam3r — APPROVED** [review]

**Pull request merged**